### PR TITLE
rgw/notification: Add support to set topic policy while topic creation

### PR DESCRIPTION
**Pull request by kchheda3**
As part of topic policy, add support to set policies on topics that gives permission to other users to access the topic. Before this change, any user could modify any topics in the system.
Also as part of setting policy, support the SetTopicAttributes() in ceph which allows setting the topic_attributes once the topic is created. Before this change user had to call CreateTopic() to change any attributes. with support  of
SetTopicAttributes(), user can now call SetTopicAttributes() to change any specific attribute without needing to pass all the attributes that were need to pass along with CreateTopic().

Fixes: https://tracker.ceph.com/issues/62727

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/common/options/rgw.yaml.in [hunk 1]
```diff
@@ -3850,3 +3850,16 @@ options:
   services:
   - rgw
   with_legacy: true
+- name: mandatory_topic_permissions
```
Comment (anthonyeleven) [inline src/common/options/rgw.yaml.in:3853]: YAML linters like the first letter of the name to be a capital.

File: src/common/options/rgw.yaml.in [hunk 2]
```diff
@@ -3850,3 +3850,16 @@ options:
...
+  type: bool
+  level: basic
+  desc: Whether to validate user permissions to access notification topics.
+  long_desc: If true, all the users will need to have a policy to access topics.
```
Comment (anthonyeleven) [inline src/common/options/rgw.yaml.in:3857]: s/the//

**Review by anthonyeleven — CHANGES_REQUESTED** [review]

File: doc/radosgw/notifications.rst [hunk 3]
```diff
@@ -179,6 +180,24 @@ Request parameters:
   default value is taken from `rgw_topic_persistency_sleep_duration`.
   providing a value overrides the global value.
   zero value mean there is no delay between retries.
+- Policy: This will control who can access the topic other than owner of the topic.
```
[the remark this answers is not in the document]
Comment (kchheda3) [reply doc/radosgw/notifications.rst:183]: Done.

File: doc/radosgw/notifications.rst [hunk 4]
```diff
@@ -179,6 +180,24 @@ Request parameters:
...
+  The policy passed needs to be json string similar to bucket policy.
```
[the remark this answers is not in the document]
Comment (kchheda3) [reply doc/radosgw/notifications.rst:184]: Done

File: doc/radosgw/notifications.rst [hunk 5]
```diff
@@ -179,6 +180,24 @@ Request parameters:
...
+  For example, one can send policy string as follows::
```
[the remark this answers is not in the document]
Comment (kchheda3) [reply doc/radosgw/notifications.rst:185]: done.

File: doc/radosgw/notifications.rst [hunk 6]
```diff
@@ -463,6 +484,61 @@ The response has the following format:
...
+
+   Action=SetTopicAttributes
+   &TopicArn=<topic-arn>&AttributeName=<attribute-name>&AttributeValue=<attribute-value>
+
+This allows to set/modify existing attributes on the specified topic.
+
+.. note::
+
+  - The Attribute name passed will either be updated or created (if not exist) with AttributeValue passed.
+  - Any un-supported attribute-name passed will result in error 400.
```
[the remark this answers is not in the document]
Comment (kchheda3) [reply doc/radosgw/notifications.rst:502]: done.

File: doc/radosgw/notifications.rst [hunk 7]
```diff
@@ -463,6 +484,61 @@ The response has the following format:
...
+
+  - push-endpoint: This is the URI of an endpoint to send push notifications to.
+  - OpaqueData: Opaque data is set in the topic configuration and added to all
+    notifications that are triggered by the topic.
+  - persistent: This indicates whether notifications to this endpoint are
+    persistent (=asynchronous) or not persistent. (This is "false" by default.)
+  - time_to_live: This will limit the time (in seconds) to retain the notifications.
+  - max_retries: This will limit the max retries before expiring notifications.
+  - retry_sleep_duration: This will control the frequency of retrying the notifications.
+  - Policy: This will control who can access the topic other than owner of the topic.
```
[the remark this answers is not in the document]
Comment (kchheda3) [reply doc/radosgw/notifications.rst:524]: Yes, coz these attributes are case-sensitive, and both OpaqueData and Policy are capitalized as per aws library.

File: doc/radosgw/notifications.rst [hunk 8]
```diff
@@ -463,6 +484,61 @@ The response has the following format:
...
+  - verify-ssl: This indicates whether the server certificate is validated by
+    the client. (This is "true" by default.)
+  - ``use-ssl``: If this is set to "true", a secure connection is used to
+   connect to the broker. (This is "false" by default.)
```
[the remark this answers is not in the document]
Comment (kchheda3) [reply doc/radosgw/notifications.rst:528]: done

File: doc/radosgw/notifications.rst [hunk 9]
```diff
@@ -463,6 +484,61 @@ The response has the following format:
...
+  - cloudevents: This indicates whether the HTTP header should contain
+    attributes according to the `S3 CloudEvents Spec`_. 
+  - amqp-exchange: The exchanges must exist and must be able to route messages
+    based on topics.
+  - amqp-ack-level: No end2end acking is required. Messages may persist in the
```
[the remark this answers is not in the document]
Comment (kchheda3) [reply doc/radosgw/notifications.rst:533]: done

File: doc/radosgw/notifications.rst [hunk 10]
```diff
@@ -463,6 +484,61 @@ The response has the following format:
...
+    broker before being delivered to their final destinations. 
+  - ``ca-location``: If this is provided and a secure connection is used, the
+    specified CA will be used instead of the default CA to authenticate the
+    broker. 
+  - mechanism: may be provided together with user/password (default: ``PLAIN``).
+  - kafka-ack-level: No end2end acking is required. Messages may persist in the
```
[the remark this answers is not in the document]
Comment (kchheda3) [reply doc/radosgw/notifications.rst:539]: done.

File: src/common/options/rgw.yaml.in (hunk 1 above)
Comment (kchheda3) [reply src/common/options/rgw.yaml.in:3853]: all the name's of the variable are lower case by default.
this is the actual variable name, so then why will it be capital ?

File: src/common/options/rgw.yaml.in (hunk 2 above)
Comment (kchheda3) [reply src/common/options/rgw.yaml.in:3857]: done

File: src/common/options/rgw.yaml.in [hunk 11]
```diff
@@ -3850,3 +3850,16 @@ options:
   services:
   - rgw
   with_legacy: true
+- name: mandatory_topic_permissions
+  type: bool
+  level: basic
+  desc: Whether to validate user permissions to access notification topics.
+  long_desc: If true, all users will need to have a policy to access topics.
+    The topic policy can be set by owner via CreateTopic() or SetTopicAttribute().
+    Following permissions can be granted snsPublish, snsGetTopicAttributes,
```
Comment (yuvalif) [inline src/common/options/rgw.yaml.in:3859]:
```suggestion
    Following permissions can be granted sns:Publish, sns:GetTopicAttributes,
```

File: src/common/options/rgw.yaml.in [hunk 12]
```diff
@@ -3850,3 +3850,16 @@ options:
...
+    snsSetTopicAttributes and snsDeleteTopic via Policy.
```
Comment (yuvalif) [inline src/common/options/rgw.yaml.in:3860]:
```suggestion
    sns:SetTopicAttributes and sns:DeleteTopic via Policy.
```

File: src/common/options/rgw.yaml.in [hunk 13]
```diff
@@ -3850,3 +3850,16 @@ options:
   services:
   - rgw
   with_legacy: true
+- name: mandatory_topic_permissions
+  type: bool
+  level: basic
+  desc: Whether to validate user permissions to access notification topics.
+  long_desc: If true, all users will need to have a policy to access topics.
```
Comment (yuvalif) [inline src/common/options/rgw.yaml.in:3857]:
```suggestion
  long_desc: If true, all users (other then the owner of the tpic) will need to have a policy to access topics.
```

File: src/common/options/rgw.yaml.in [hunk 14]
```diff
@@ -3850,3 +3850,16 @@ options:
...
+    The topic policy can be set by owner via CreateTopic() or SetTopicAttribute().
+    Following permissions can be granted snsPublish, snsGetTopicAttributes,
+    snsSetTopicAttributes and snsDeleteTopic via Policy.
+  default: false
```
Comment (yuvalif) [inline src/common/options/rgw.yaml.in:3861]: please add: "note that even if set to "false" topics will follow the policies set on them"

File: src/rgw/rgw_pubsub.cc [hunk 15]
```diff
@@ -337,6 +337,7 @@ void rgw_pubsub_topic::dump(Formatter *f) const
   encode_json("dest", dest, f);
   encode_json("arn", arn, f);
   encode_json("opaqueData", opaque_data, f);
+  encode_json("Policy", policy_text, f);
```
Comment (yuvalif) [inline src/rgw/rgw_pubsub.cc:340]: i think that JSON should be `camelCase`, so we should change to "policy"

File: src/rgw/rgw_rest_pubsub.cc [hunk 16]
```diff
@@ -76,25 +123,41 @@ class RGWPSCreateTopicOp : public RGWOp {
...
+  std::string policy_text;
+
   int get_params() {
     topic_name = s->info.args.get("Name");
     if (topic_name.empty()) {
       ldpp_dout(this, 1) << "CreateTopic Action 'Name' argument is missing" << dendl;
       return -EINVAL;
     }
 
+    // Remove the args that are parsed, so the push_endpoint_args only contains
```
Comment (yuvalif) [inline src/rgw/rgw_rest_pubsub.cc:135]: why is this needed?

File: src/rgw/rgw_rest_pubsub.cc [hunk 17]
```diff
@@ -163,14 +242,18 @@ class RGWPSCreateTopicOp : public RGWOp {
 };
 
 void RGWPSCreateTopicOp::execute(optional_yield y) {
-  op_ret = get_params();
-  if (op_ret < 0) {
-    return;
+  if (!dest.push_endpoint.empty() && dest.persistent) {
```
Comment (yuvalif) [inline src/rgw/rgw_rest_pubsub.cc:245]: great! this is the right place

File: src/rgw/rgw_rest_pubsub.cc [hunk 18]
```diff
@@ -125,9 +181,32 @@ class RGWPSCreateTopicOp : public RGWOp {
...
+    if (ret == -ENOENT) {
+      // topic not present
+      return 0;
+    }
+    if (ret == 0) {
+      if (result.user == s->owner.get_id() ||
+          !s->cct->_conf->mandatory_topic_permissions) {
+        return 0;
+      }
+      ldpp_dout(this, -1) << "failed to create topic '" << topic_name
```
Comment (yuvalif) [inline src/rgw/rgw_rest_pubsub.cc:202]: please avoid "-1" debug level. these cannot be turned off

File: src/rgw/rgw_pubsub.cc (hunk 15 above)
Comment (kchheda3) [reply src/rgw/rgw_pubsub.cc:340]: Done.

**Comment by kchheda3** [comment]
> please avoid "-1" debug level. these cannot be turned off 

My understanding was error's are "-1" and should not be allowed to supress so -1 ?

**Comment by kchheda3** [comment]
> why is this needed? 

so currently without removing the args, all the attributes get set into the push_endpoints_args and while listing the topic all the params show twice, once against their struct and once inside the push_endpoint_args.
for example, push_endpoint value is shown in dest.push_endpoint as well as dest.push_endpoint_args
we could exclude them explicitly below or just remove them here.

**Comment by kchheda3** [comment]
> ```suggestion
    Following permissions can be granted sns:Publish, sns:GetTopicAttributes,
``` 

":" is a key word for the yaml and it throws error when we keep `sns":"....`
so i removed the ":"

File: src/common/options/rgw.yaml.in (hunk 12 above)
Comment (kchheda3) [reply src/common/options/rgw.yaml.in:3860]: > 
":" is a key word for the yaml and it throws error when we keep `sns":"....`
so i removed the ":"

File: src/common/options/rgw.yaml.in (hunk 11 above)
Comment (kchheda3) [reply src/common/options/rgw.yaml.in:3859]: > 
":" is a key word for the yaml and it throws error when we keep `sns":"....`
so i removed the ":"

File: src/rgw/rgw_rest_pubsub.cc (hunk 16 above)
Comment (kchheda3) [reply src/rgw/rgw_rest_pubsub.cc:135]: > why is this needed?

so currently without removing the args, all the attributes get set into the push_endpoints_args and while listing the topic all the params show twice, once against their struct and once inside the push_endpoint_args.
for example, push_endpoint value is shown in dest.push_endpoint as well as dest.push_endpoint_args
we could exclude them explicitly below or just remove them here.

File: src/rgw/rgw_rest_pubsub.cc (hunk 18 above)
Comment (kchheda3) [reply src/rgw/rgw_rest_pubsub.cc:202]: > please avoid "-1" debug level. these cannot be turned off

My understanding was error's are "-1" and should not be allowed to supress so -1 ?

File: src/rgw/rgw_rest_pubsub.cc (hunk 17 above)
Comment (kchheda3) [reply src/rgw/rgw_rest_pubsub.cc:245]: > great! this is the right place

yep this and same for the DeleteTopic, we were deleting in get_params()

File: src/common/options/rgw.yaml.in (hunk 14 above)
Comment (kchheda3) [reply src/common/options/rgw.yaml.in:3861]: Done

File: src/common/options/rgw.yaml.in (hunk 12 above)
Comment (yuvalif) [reply src/common/options/rgw.yaml.in:3861]: i think you can put the test in quotes, e.g. `"sns:DeleteTopics"`
or put the pipe sign before the text. e.g.
```
  long_desc: |
    If true, all users...
```

File: src/rgw/rgw_rest_pubsub.cc (hunk 18 above)
Comment (yuvalif) [reply src/rgw/rgw_rest_pubsub.cc:202]: this is a wider discussion, but, IMO, we always need a way to turn off debug logs. otherwise these errors can cause denial of service.

File: src/common/options/rgw.yaml.in (hunk 12 above)
Comment (kchheda3) [reply src/common/options/rgw.yaml.in:3861]: > i think you can put the test in quotes, e.g. `"sns:DeleteTopics"` or put the pipe sign before the text. e.g.
> 
> ```
>   long_desc: |
>     If true, all users...
> ```

Thanks done.

File: src/common/options/rgw.yaml.in (hunk 11 above)
Comment (kchheda3) [reply src/common/options/rgw.yaml.in]: done

File: src/rgw/rgw_rest_pubsub.cc (hunk 18 above)
Comment (yuvalif) [reply src/rgw/rgw_rest_pubsub.cc:202]: high rate of error logs can slow down the system. it should be possible to turn them off if the user decide that the harm caused by this slowdown is greater than the value these logs provide

File: doc/radosgw/notifications.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (yuvalif) [reply doc/radosgw/notifications.rst:524]: @anthonyeleven can you please re-review?

File: src/rgw/rgw_rest_pubsub.cc (hunk 18 above)
Comment (kchheda3) [reply src/rgw/rgw_rest_pubsub.cc:202]: > > this is a wider discussion, but, IMO, we always need a way to turn off debug logs. otherwise these errors can cause denial of service.
> 
> denial of service ?? not sure if i understand this ? not able to turn of debug/error messages can lead to denial of service ? I mean any S3 operation that is failing, should have an option to turn of that error log IMO is little weird ?

@yuvalif as discussed during the rgw weekly, it was recommend to avoid -1 specially during the http_request that are send, so will remove change the log level to 1 from -1.
thanks

**Review by yuvalif — APPROVED** [review]

File: src/rgw/rgw_rest_pubsub.cc [hunk 19]
```diff
@@ -463,6 +743,30 @@ void RGWPSDeleteTopicOp::execute(optional_yield y) {
...
+    op_ret = verify_topic_owner_or_policy(
+        s, result, driver->get_zone()->get_zonegroup().get_name(),
+        rgw::IAM::snsDeleteTopic);
+    if (op_ret != 0) {
+      ldpp_dout(this, 1) << "failed to remove topic '" << topic_name
+                         << "' topic owned by other user" << dendl;
+      return;
+    }
+  } else {
+    ldpp_dout(this, 1) << "failed to fetch topic '" << topic_name << dendl;
```
Comment (yuvalif) [inline src/rgw/rgw_rest_pubsub.cc:758]: please change `op_ret` to `0`
currently the `test_ps_s3_notification_errors_on_master` test is failing since it gets `http status=404`.
the expected behavior is that if we delete an already deleted topic or a non-existent topic we should get `http statu=200`

Comment (kchheda3) [reply src/rgw/rgw_rest_pubsub.cc:758]: how about changing the test then, it makes sense to return 404 for a topic that did not exist ??

Comment (yuvalif) [reply src/rgw/rgw_rest_pubsub.cc:758]: the existing clients are not expceting a 404 on deleting an unknown entity (topic, object, bucket etc.)
we should keep the existing behavior, and return success in this case

Comment (kchheda3) [reply src/rgw/rgw_rest_pubsub.cc:758]: i just double checked to delete a non-existent bucket and i get 404
` s3cmd ls --verbose
2023-10-19 14:25  s3://testbkt

>> s3cmd rb s3://test --verbose
ERROR: S3 error: 404 (NoSuchBucket)
`

Comment (kchheda3) [reply src/rgw/rgw_rest_pubsub.cc:758]: > i just double checked to delete a non-existent bucket and i get 404 ` s3cmd ls --verbose 2023-10-19 14:25 s3://testbkt
> 
> > > s3cmd rb s3://test --verbose
> > > ERROR: S3 error: 404 (NoSuchBucket)
> > > `

as can be seen here in the code --> https://github.com/ceph/ceph/blob/main/src/rgw/rgw_op.cc#L3527-L3530

Comment (yuvalif) [reply src/rgw/rgw_rest_pubsub.cc:758]: you are right regarding bucket deletion, but there are other places where we consider such deletions as success.
look for code that looks like that
```
if (op_ret == -ENOENT) {
  op_ret = 0;
}
```
regardles, the existing behavior regarding topic deletion is to consider such deletion as success.

File: doc/radosgw/notifications.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (yuvalif) [reply doc/radosgw/notifications.rst:525]: @anthonyeleven ping

File: src/rgw/rgw_rest_pubsub.cc (hunk 19 above)
Comment (kchheda3) [reply src/rgw/rgw_rest_pubsub.cc]: > you are right regarding bucket deletion, but there are other places where we consider such deletions as success. look for code that looks like that
> 
> ```
> if (op_ret == -ENOENT) {
>   op_ret = 0;
> }
> ```
> 
> regardles, the existing behavior regarding topic deletion is to consider such deletion as success.

done

**Review by yuvalif — APPROVED** [review]

**Comment by yuvalif** [comment]
teuthology passing: http://pulpito.front.sepia.ceph.com/yuvalif-2023-10-29_14:41:55-rgw:notifications-wip-topic-policy-distro-default-smithi/

File: doc/radosgw/notifications.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (yuvalif) [reply doc/radosgw/notifications.rst:525]: @anthonyeleven can you please re-review?

**Review by anthonyeleven — APPROVED** [review]
Approved docs only

File: doc/radosgw/notifications.rst [hunk 20]
```diff
@@ -159,6 +159,7 @@ updating, use the name of an existing topic and different endpoint values).
    [&Attributes.entry.12.key=time_to_live&Attributes.entry.12.value=<seconds to live>]
    [&Attributes.entry.13.key=max_retries&Attributes.entry.13.value=<retries number>]
    [&Attributes.entry.14.key=retry_sleep_duration&Attributes.entry.14.value=<sleep seconds>]
+   [&Attributes.entry.14.key=Policy&Attributes.entry.14.value=<policy-JSON-string>]
```
Comment (jzhu116-bloomberg) [inline doc/radosgw/notifications.rst:162]: There are 2 `Attributes.entry.14`. Is it on purpose?

Comment (kchheda3) [reply doc/radosgw/notifications.rst:162]: > There are 2 `Attributes.entry.14`. Is it on purpose?

typo

Comment (cbodley) [reply doc/radosgw/notifications.rst:162]: thanks, see https://github.com/ceph/ceph/pull/55403
